File: src/main/java/edu/missouristate/mars/mips/instructions/syscalls/SyscallExit2.java
package edu.missouristate.mars.mips.instructions.syscalls;

import edu.missouristate.mars.util.*;
import edu.missouristate.mars.*;
import edu.missouristate.mars.mips.hardware.*;


/**
 * Service to exit the MIPS program with return value given in $a0.  Ignored if running from GUI.
 */

public class SyscallExit2 extends AbstractSyscall {
    /**
     * Build an instance of the Exit2 syscall.  Default service number
     * is 17 and name is "Exit2".
     */
    public SyscallExit2() {
        super(17, "Exit2");
    }

    /**
     * Performs syscall function to exit the MIPS program with return value given in $a0.
     * If running in command mode, MARS will exit with that value.  If running under GUI,
     * return value is ignored.
     */
    public void simulate(ProgramStatement statement) throws ProcessingException {
        if (Globals.getGui() == null) {
            Globals.exitCode = RegisterFile.getValue(4);
        }
        throw new ProcessingException(); // empty error list
    }
}

File: src/main/java/edu/missouristate/mars/mips/instructions/syscalls/SyscallExit2.kt
/*
 * Copyright (c) 2003-2024, Pete Sanderson and Kenneth Vollmar
 * Copyright (c) 2024-present, Nicholas Hubbard
 *
 * Originally developed by Pete Sanderson (psanderson@otterbein.edu) and Kenneth Vollmar (kenvollmar@missouristate.edu)
 * Maintained by Nicholas Hubbard (nhubbard@users.noreply.github.com)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
 * documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * 1. The above copyright notice and this permission notice shall be included in all copies or substantial portions of
 *    the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
 * WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
 * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package edu.missouristate.mars.mips.instructions.syscalls

import edu.missouristate.mars.Globals.exitCode
import edu.missouristate.mars.Globals.gui
import edu.missouristate.mars.ProcessingException
import edu.missouristate.mars.ProgramStatement
import edu.missouristate.mars.mips.hardware.RegisterFile.getValue

/**
 * Service to exit the MIPS program with return value given in $a0.  Ignored if running from GUI.
 */
class SyscallExit2 : AbstractSyscall(17, "Exit2") {
    /**
     * Performs syscall function to exit the MIPS program with return value given in $a0.
     * If running in command mode, MARS will exit with that value.  If running under GUI,
     * return value is ignored.
     */
    @Throws(ProcessingException::class)
    override fun simulate(statement: ProgramStatement) {
        if (gui == null) exitCode = getValue(4)
        throw ProcessingException() // empty error list
    }
}